File: android/app/src/main/java/com/bkes994408/expensetracker/telemetry/Telemetry.kt
package com.bkes994408.expensetracker.telemetry

import android.os.SystemClock
import android.util.Log

enum class AnalyticsEvent {
    APP_LAUNCHED,
    APP_STARTUP_MEASURED,
    EXPENSE_ADDED,
    EXPENSE_ADD_INVALID,
    UNCAUGHT_EXCEPTION,
}

interface AnalyticsService {
    fun track(event: AnalyticsEvent, metadata: Map<String, String> = emptyMap())
}

interface CrashReporter {
    fun install()
    fun record(throwable: Throwable, fatal: Boolean, metadata: Map<String, String> = emptyMap())
}

class LogcatAnalyticsService : AnalyticsService {
    override fun track(event: AnalyticsEvent, metadata: Map<String, String>) {
        Log.i("ExpenseTelemetry", "event=${event.name} metadata=$metadata")
    }
}

class LogcatCrashReporter : CrashReporter {
    private val previousHandler = Thread.getDefaultUncaughtExceptionHandler()

    override fun install() {
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            Telemetry.record(throwable, fatal = true, metadata = mapOf("thread" to thread.name))
            previousHandler?.uncaughtException(thread, throwable)
        }
    }

    override fun record(throwable: Throwable, fatal: Boolean, metadata: Map<String, String>) {
        Log.e("ExpenseTelemetry", "fatal=$fatal metadata=$metadata", throwable)
    }
}

object Telemetry {
    private val appStartElapsedRealtimeMs = SystemClock.elapsedRealtime()
    private val analytics: AnalyticsService by lazy { LogcatAnalyticsService() }
    private val crashReporter: CrashReporter by lazy { LogcatCrashReporter() }

    fun install() {
        crashReporter.install()
        track(AnalyticsEvent.APP_LAUNCHED)
    }

    fun track(event: AnalyticsEvent, metadata: Map<String, String> = emptyMap()) {
        analytics.track(event, metadata)
    }

    fun record(throwable: Throwable, fatal: Boolean = false, metadata: Map<String, String> = emptyMap()) {
        if (fatal) {
            track(AnalyticsEvent.UNCAUGHT_EXCEPTION, metadata)
        }
        crashReporter.record(throwable, fatal, metadata)
    }

    fun markFirstFrameDrawn() {
        val elapsedMs = SystemClock.elapsedRealtime() - appStartElapsedRealtimeMs
        track(AnalyticsEvent.APP_STARTUP_MEASURED, mapOf("elapsedMs" to elapsedMs.toString()))
    }
}

File: ios/ExpenseTracker/Telemetry/Telemetry.swift
import Foundation
import os

enum AnalyticsEvent: String {
    case appLaunched = "app_launched"
    case expenseAdded = "expense_added"
    case expenseAddInvalid = "expense_add_invalid"
    case expenseDeleted = "expense_deleted"
    case csvExported = "csv_exported"
    case csvExportFailed = "csv_export_failed"
    case proPaywallViewed = "pro_paywall_viewed"
    case proPaywallCtaTapped = "pro_paywall_cta_tapped"
}

protocol AnalyticsService {
    func track(event: AnalyticsEvent, metadata: [String: String])
}

protocol CrashReportingService {
    func install()
    func record(error: Error, fatal: Bool, metadata: [String: String])
    func recordFatal(message: String, metadata: [String: String])
}

final class ConsoleAnalyticsService: AnalyticsService {
    private let logger = Logger(subsystem: "com.bkes994408.expensetracker", category: "analytics")

    func track(event: AnalyticsEvent, metadata: [String: String]) {
        logger.log("event=\(event.rawValue, privacy: .public) metadata=\(String(describing: metadata), privacy: .public)")
    }
}

final class ConsoleCrashReporter: CrashReportingService {
    private let logger = Logger(subsystem: "com.bkes994408.expensetracker", category: "crash")

    func install() {
        NSSetUncaughtExceptionHandler { exception in
            Telemetry.shared.recordFatal(
                message: "\(exception.name.rawValue): \(exception.reason ?? "")",
                metadata: ["source": "uncaught_exception"]
            )
        }
    }

    func record(error: Error, fatal: Bool, metadata: [String: String]) {
        logger.error("fatal=\(fatal, privacy: .public) error=\(error.localizedDescription, privacy: .public) metadata=\(String(describing: metadata), privacy: .public)")
    }

    func recordFatal(message: String, metadata: [String: String]) {
        logger.fault("fatal=1 message=\(message, privacy: .public) metadata=\(String(describing: metadata), privacy: .public)")
    }
}

final class Telemetry {
    static let shared = Telemetry()

    private let analytics: AnalyticsService = ConsoleAnalyticsService()
    private let crashReporter: CrashReportingService = ConsoleCrashReporter()

    private init() {}

    func install() {
        crashReporter.install()
        track(.appLaunched)
    }

    func track(_ event: AnalyticsEvent, metadata: [String: String] = [:]) {
        analytics.track(event: event, metadata: metadata)
    }

    func record(error: Error, fatal: Bool = false, metadata: [String: String] = [:]) {
        crashReporter.record(error: error, fatal: fatal, metadata: metadata)
    }

    func recordFatal(message: String, metadata: [String: String] = [:]) {
        crashReporter.recordFatal(message: message, metadata: metadata)
    }
}
